### ISSUE TITLE
The Model Performance Mismatch Problem (and what to do about it)

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#489]
The Model Performance Mismatch Problem (and what to do about it)<br>
<br>
https://ift.tt/2qEVAmh<br>
<br>
Jason Brownlee<br>
<br>
<h4>What To Do If Model Test Results Are Worse than Training.</h4>
<p>The procedure when evaluating machine learning models is to fit and evaluate them on training data, then verify that the model has good skill on a held-back test dataset.</p>
<p>Often, you will get a very promising performance when evaluating the model on the training dataset and poor performance when evaluating the model on the test set.</p>
<p>In this post, you will discover techniques and issues to consider when you encounter this common problem.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>The problem of model performance mismatch that may occur when evaluating machine learning algorithms.</li>
<li>The causes of overfitting, under-representative data samples, and stochastic algorithms.</li>
<li>Ways to harden your test harness to avoid the problem in the first place.</li>
</ul>
<p>This post was based on a reader question; thanks! Keep the questions coming!</p>
<p>Let&rsquo;s get started.</p>
<div class="wp-caption">
<img src="https://machinelearningmastery.com/wp-content/uploads/2018/04/The-Model-Performance-Mismatch-Problem-and-what-to-do-about-it.jpg" alt="The Model Performance Mismatch Problem (and what to do about it)" width="640" class="wp-image-5121" height="427">
<p class="wp-caption-text">The Model Performance Mismatch Problem (and what to do about it)<br>
Photo by <a href="https://www.flickr.com/photos/arapaoaphotography/33705234080/">Arapaoa Moffat</a>, some rights reserved.</p>
</div>
<h2>Overview</h2>
<p>This post is divided into 4 parts; they are:</p>
<ol>
<li>Model Evaluation</li>
<li>Model Performance Mismatch</li>
<li>Possible Causes and Remedies</li>
<li>More Robust Test Harness</li>
</ol>
<h2>Model Evaluation</h2>
<p>When developing a model for a predictive modeling problem, you need a test harness.</p>
<p>The test harness defines how the sample of data from the domain will be used to evaluate and compare candidate models for your predictive modeling problem.</p>
<p>There are many ways to structure a test harness, and no single best way for all projects.</p>
<p>One popular approach is to use a portion of data for fitting and tuning the model and a portion for providing an objective estimate of the skill of the tuned model on out-of-sample data.</p>
<p>The data sample is split into a training and test dataset. The model is evaluated on the training dataset using a resampling method such as k-fold cross-validation, and the set itself may be further divided into a validation dataset used to tune the hyperparameters of the model.</p>
<p>The test set is held back and used to evaluate and compare tuned models.</p>
<p>For more on training, validation, and test sets, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/difference-test-validation-datasets/">What is the Difference Between Test and Validation Datasets?</a></li>
</ul>
<h2>Model Performance Mismatch</h2>
<p>The resampling method will give you an estimate of the skill of your model on unseen data by using the training dataset.</p>
<p>The test dataset provides a second data point and ideally an objective idea of how well the model is expected to perform, corroborating the estimated model skill.</p>
<p>What if the estimate of model skill on the training dataset does not match the skill of the model on the test dataset?</p>
<p>The scores will not match in general. We do expect some differences because some small overfitting of the training dataset is inevitable given hyperparameter tuning, making the training scores optimistic.</p>
<p>But what if the difference is worryingly large?</p>
<ul>
<li>Which score do you trust?</li>
<li>Can you still compare models using the test dataset?</li>
<li>Is the model tuning process invalidated?</li>
</ul>
<p>It is a challenging and very common situation in applied machine learning.</p>
<p>We can call this concern the &ldquo;<em>model performance mismatch</em>&rdquo; problem.</p>
<p><strong>Note</strong>: ideas of &ldquo;<em>large differences</em>&rdquo; in model performance are relative to your chosen performance measures, datasets, and models. We cannot talk objectively about differences in general, only relative differences that you must interpret yourself.</p>
<h2>Possible Causes and Remedies</h2>
<p>There are many possible causes for the model performance mismatch problem.</p>
<p>Ultimately, your goal is to have a test harness that you know allows you to make good decisions regarding which model and model configuration to use as a final model.</p>
<p>In this section, we will look at some possible causes, diagnostics, and techniques you can use to investigate the problem.</p>
<p>Let&rsquo;s look at three main areas: model overfitting, the quality of the data sample, and the stochastic nature of the learning algorithm.</p>
<h3>1. Model Overfitting</h3>
<p>Perhaps the most common cause is that you have overfit the training data.</p>
<p>You have hit upon a model, a set of model hyperparameters, a view of the data, or a combination of these elements and more that just so happens to give a good skill estimate on the training dataset.</p>
<p>The use of k-fold cross-validation will help to some degree. The use of tuning models with a separate dataset too will help. Nevertheless, it is possible to keep pushing and overfit on the training dataset.</p>
<p>If this is the case, the test skill may be more representative of the true skill of the chosen model and configuration.</p>
<p>One simple (but not easy) way to diagnose whether you have overfit the training dataset, is to get another data point on model skill. Evaluate the chosen model on another set of data. For example, some ideas to try include:</p>
<ul>
<li>Try a k-fold cross-validation evaluation of the model on the test dataset.</li>
<li>Try a fit of the model on the training dataset and an evaluation on the test and a new data sample.</li>
</ul>
<p>If you&rsquo;re overfit, you have options.</p>
<ul>
<li>Perhaps you can scrap your current training dataset and collect a new training dataset.</li>
<li>Perhaps you can re-split your sample into train/test in a softer approach to getting a new training dataset.</li>
</ul>
<p>I would suggest that the results that you have obtained to-date are suspect and should be re-considered. Especially those where you may have spent too long tuning.</p>
<p>Overfitting may be the ultimate cause for the discrepancy in model scores, though it may not be the area to attack first.</p>
<h3>2. Unrepresentative Data Sample</h3>
<p>It is possible that your training or test datasets are an unrepresentative sample of data from the domain.</p>
<p>This means that the sample size is too small or the examples in the sample do not effectively &ldquo;cover&rdquo; the cases observed in the broader domain.</p>
<p>This can be obvious to spot if you see noisy model performance results. For example:</p>
<ul>
<li>A large variance on cross-validation scores.</li>
<li>A large variance on similar model types on the test dataset.</li>
</ul>
<p>In addition, you will see the discrepancy between train and test scores.</p>
<p>Another good second test is to check summary statistics for each variable on the train and test sets, and ideally on the cross-validation folds. You are looking for a large variance in sample means and standard deviation.</p>
<p>The remedy is often to get a larger and more representative sample of data from the domain. Alternately, to use more discriminating methods in preparing the data sample and splits. Think stratified k-fold cross validation, but applied to input variables in an attempt to maintain population means and standard deviations for real-valued variables in addition to the distribution of categorical variables.</p>
<p>Often when I see overfitting on a project, it is because the test harness is not as robust as it should be, not because of hill climbing the test dataset.</p>
<h3>3. Stochastic Algorithm</h3>
<p>It is possible that you are seeing a discrepancy in model scores because of the stochastic nature of the algorithm.</p>
<p>Many machine learning algorithms involve a stochastic component. For example, the random initial weights in a neural network, the shuffling of data and in turn the gradient updates in stochastic gradient descent, and much more.</p>
<p>This means, that each time the same algorithm is run on the same data, different sequences of random numbers are used and, in turn, a different model with different skill will result.</p>
<p>You can learn more about this in the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/randomness-in-machine-learning/">Embrace Randomness in Machine Learning</a></li>
</ul>
<p>This issue can be seen by the variance in model skill scores from cross-validation, much like having an unrepresentative data sample.</p>
<p>The difference here is that the variance can be cleared up by repeating the model evaluation process, e.g. cross-validation, in order to control for the randomness in training the model.</p>
<p>This is often called the multiple repeats k-fold cross-validation and is used for neural networks and stochastic optimization algorithms, when resources permit.</p>
<p>I have more on this approach to evaluating models in the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/evaluate-skill-deep-learning-models/">How to Evaluate the Skill of Deep Learning Models</a></li>
</ul>
<h2>More Robust Test Harness</h2>
<p>A lot of these problems can be addressed early by designing a robust test harness and then gathering evidence to demonstrate that indeed your test harness is robust.</p>
<p>This might include running experiments before you get started evaluating models for real. Experiments such:</p>
<ul>
<li>A sensitivity analysis of train/test splits.</li>
<li>A sensitivity analysis of k values for cross-validation.</li>
<li>A sensitivity analysis of a given model&rsquo;s behavior.</li>
<li>A sensitivity analysis on the number of repeats.</li>
</ul>
<p>On this last point, see the post:</p>
<ul>
<li><a href="https://machinelearningmastery.com/estimate-number-experiment-repeats-stochastic-machine-learning-algorithms/">Estimate the Number of Experiment Repeats for Stochastic Machine Learning Algorithms</a></li>
</ul>
<p>You are looking for:</p>
<ul>
<li>Low variance and consistent mean in evaluation scores between tests in a cross-validation.</li>
<li>Correlated population means between model scores on train and test sets.</li>
</ul>
<p>Use statistical tools like standard error and significance tests if needed.</p>
<p>Use a modern and un-tuned model that performs well in general for such testing, such as random forest.</p>
<ul>
<li>If you discover a difference in skill scores between training and test sets, and it is consistent, that may be fine. You know what to expect.</li>
<li>If you measure a variance in mean skill scores within a given test, you have error bars you can use to interpret the results.</li>
</ul>
<p>I would go so far as to say that without a robust test harness, the results you achieve will be a mess. You will not be able to effectively interpret them. There will be an element of risk (or fraud, if you&rsquo;re an academic) in the presentation of the outcomes from a fragile test harness. And reproducibility/robustness is a massive problem in numerical fields like applied machine learning.</p>
<p>Finally, avoid using the test dataset too much. Once you have strong evidence that your harness is robust, do not touch the test dataset until it comes time for final model selection.</p>
<h2>Further Reading</h2>
<p>This section provides more resources on the topic if you are looking to go deeper.</p>
<ul>
<li><a href="https://machinelearningmastery.com/difference-test-validation-datasets/">What is the Difference Between Test and Validation Datasets?</a></li>
<li><a href="https://machinelearningmastery.com/randomness-in-machine-learning/">Embrace Randomness in Machine Learning</a></li>
<li><a href="https://machinelearningmastery.com/estimate-number-experiment-repeats-stochastic-machine-learning-algorithms/">Estimate the Number of Experiment Repeats for Stochastic Machine Learning Algorithms</a></li>
<li><a href="https://machinelearningmastery.com/evaluate-skill-deep-learning-models/">How to Evaluate the Skill of Deep Learning Models</a></li>
</ul>
<h2>Summary</h2>
<p>In this post, you discovered the model performance mismatch problem where model performance differs greatly between training and test sets, and techniques to diagnose and address the issue.</p>
<p>Specifically, you learned:</p>
<ul>
<li>The problem of model performance mismatch that may occur when evaluating machine learning algorithms.</li>
<li>The causes of overfitting, under-representative data samples, and stochastic algorithms.</li>
<li>Ways to harden your test harness to avoid the problem in the first place.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
<p>The post <a rel="nofollow" href="https://machinelearningmastery.com/the-model-performance-mismatch-problem/">The Model Performance Mismatch Problem (and what to do about it)</a> appeared first on <a rel="nofollow" href="https://machinelearningmastery.com">Machine Learning Mastery</a>.</p>
<br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
April 18, 2018 at 03:03AM
